feat(NPU): add UB Manager for auto tiling strategy management

noemotiovon opened the pull request:
### Background and Motivation

When developing Ascend NPU operators, we frequently encounter compilation failures caused by UB (Unified Buffer) overflow. During compilation, Triton kernels check UB usage, and if it exceeds the capacity, an error is raised:
 `MLIRCompilationError: ub overflow`.

To address this issue, developers usually have to:

- Manually adjust block sizes
- Tune separately for different NPU models and input sizes
- Repeatedly trade off between performance and UB safety

This process is tedious, error-prone, and difficult to cover all scenarios.

------

### Solution

We implemented a **UB Manager** that provides:

- **Automatic UB capacity detection**: retrieved from device properties or environment variables
- **Best-practice-based tiling strategies**: dynamically compute safe block sizes based on UB capacity and operator parameters
- **Unified strategy registration system**: supports both fixed and conditional strategies, making it easy to extend
- **Integration with GEGLU and ROPE**: automatically handles UB constraints and prevents overflow

------

### Core Features

#### Automatic Capacity Detection

- Supports three sources: environment variables, device properties, and model default values
- Supports multiple Ascend models such as Ascend 910B1 / 910B4

#### Dynamic Strategy Computation

- **GEGLU**: computes a safe block size based on `n_cols` and `dtype_size`
- **ROPE**: computes `BLOCK_Q` and `BLOCK_K` based on `pad_n_q_head`, `pad_n_kv_head`, and `pad_hd`
- Uses an **80% safety margin** to balance performance and safety

#### Easy Extensibility

- Simple interface: adding a new kernel strategy only requires registering a function
- Supports parameterized strategies to adapt to different input sizes

------

### Implementation Details

- Added `ub_manager.py`: the core UB management class
- Updated `geglu.py` and `rope.py`: integrated UB-aware implementations

------

### Testing

Verified on Ascend NPU 910B4:

- GEGLU forward and backward pass tests
- ROPE forward and backward pass tests
- Works correctly across different input sizes and data types
- Hardware Type: <Ascend910B4>
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

TianHao324 commented on src/liger_kernel/ops/backends/_ascend/ops/rope.py:
```diff
@@ -0,0 +1,228 @@
+    else:
+        # Fallback to conservative defaults
+        BLOCK_Q = triton.next_power_of_2(pad_n_q_head)
+        BLOCK_K = triton.next_power_of_2(pad_n_kv_head)
```
Shouldn't this be the same logic as the corresponding one in `rope_back`?
```
BLOCK_Q = min(32, triton.next_power_of_2(pad_n_q_head))
BLOCK_K = min(32, triton.next_power_of_2(pad_n_kv_head))
```

Tcc0403 reviewed (commented):
This is cool! Overall lgtm, just some questions about caching and parameter normalization.

Regarding tolerances in geglu test, I think #986 doesn't comletely fix the numerical issue, it still needs further investigation. It's fine to relax the tolerance as you need.

noemotiovon commented on src/liger_kernel/ops/backends/_ascend/ub_manager.py:
```diff
@@ -0,0 +1,365 @@
+                            )
+                    else:
+                        # Generic: convert dict values to tuple (sorted for consistency)
+                        normalized_params = tuple(sorted(key_params.values()))
```
You're right. The previous implementation was incorrect.
We don't need to convert dict to tuple in `get_tiling_strategy`. `_geglu_default_strategy` and `_rope_default_strategy` already accept both dict and tuple, and handle both internally.
I've simplified the code by removing the conversion logic and passing `key_params` directly to the strategy functions. Parameter parsing is now handled inside each strategy function, which:
* Makes `get_tiling_strategy` only responsible for finding and calling strategies
* Lets strategy functions handle their own parameter formats
This keeps responsibilities clear and makes it easier to extend. When adding new strategy functions, we only need to handle parameter formats inside the function, without modifying `get_tiling_strategy`.

noemotiovon commented on src/liger_kernel/ops/backends/_ascend/ub_manager.py:
```diff
@@ -0,0 +1,365 @@
+}
+
+
+def _geglu_default_strategy(key_params: Optional[Union[Tuple, Dict]]) -> Optional[Tuple]:
```
Good idea! I've implemented an LRU cache to cache the strategy results. The cache has a configurable size limit (default 128 entries) and uses OrderedDict for LRU eviction. This should help avoid recomputing strategies for the same parameters.

noemotiovon commented on src/liger_kernel/ops/backends/_ascend/ops/geglu.py:
```diff
@@ -0,0 +1,210 @@
+
+    # Get tiling strategy from best practices
+    dtype_size = 4 if a.dtype == torch.float32 else 2
+    strategy = get_tiling_strategy("geglu_forward", (n_cols, dtype_size))
```
Done.

noemotiovon commented:
Hi @Tcc0403,
Thank you so much for taking the time to review this, especially during the Christmas holidays — really appreciate it. Merry Christmas! 🎄

Your suggestions were very helpful. I’ve updated the code accordingly, added caching, and improved the overall functionality.

Tcc0403 commented:
> Thank you so much for taking the time to review this, especially during the Christmas holidays — really appreciate it. Merry Christmas! 🎄

My pleasure! Thank you for the contribution during holidays as well. Merry Christmas!

Just one more thing I forgot to mention, we should include your ub manager document for future contributors. You can put it under `src/liger_kernel/ops/backends/_ascend` directly without having to change anything since it's well-written already. After adding the document, I think we are good to merge this PR!

noemotiovon commented:
@Tcc0403, Thank you so much! I really appreciate it 😊
I actually wanted to add the document earlier as well, but wasn’t sure where the most appropriate place would be, haha. Thanks a lot for the guidance.
I’ll put it under `src/liger_kernel/ops/backends/_ascend` as suggested, and I’ll also expand the document based on the existing content to include the cache-related details, so it can be as comprehensive as possible for future contributors.
Thanks again for the helpful feedback!

noemotiovon commented on src/liger_kernel/ops/backends/_ascend/ub_manager.py:
```diff
@@ -0,0 +1,376 @@
+}
+
+
+def _geglu_default_strategy(key_params: Optional[Union[Tuple, Dict]]) -> Optional[Tuple]:
```
I agree with your point. My current thinking is that the parameters can actually be divided into three categories:
1. data related to the dimensions that need tiling,
2. data related to computing the tiling values, and
3. UB-related information.

Ultimately, the goal is to derive a tiling result. As for how to abstract tensor shapes, I haven’t fully figured that out yet. It probably needs to evolve alongside the introduction of more operators, so we can have more references and arrive at a more general abstraction.
I’ll first refactor things based on this line of thinking, and once that’s done, we can discuss it together again. Also, thank you very much for taking the time out of your busy schedule to discuss this with us! 😊

Tcc0403 commented on src/liger_kernel/ops/backends/_ascend/ascend-ub-manager-design.md:
```diff
@@ -0,0 +1,371 @@
+  - All elements are multiplied together to get the final unit_param
+  - For GEGLU: `()` (empty tuple, unit_param = 1)
+  - For ROPE: `(pad_hd,)` (unit_param = pad_hd)
+  - For kernels with multiple factors: `(factor1, factor2, ...)` (unit_param = factor1 * factor2 * ...)
```
In my opinion, `unit_params` is a little bit ambiguous. It's hard to tell the usage of this parameter without tracking source code.

Combining the documnent and the source code, my understanding is that `unit_params` is the fixed dimensions paired with `tiling_dims`, the flexible dimensions. However, changing it to `fixed_dims` wouldn't help much since both of them are partial dimensions of tiles, so it would also need to rename `tiling_dims` to  `flexible_dims` to be semantically correct, just me being nitpicking. It adds another level of cognitive burden and they are not self-explained without treating two parameters as a pair.

Another concern is that we implicitly assume all `tiling_dims` sharing same `unit_params` factors. That being said, this issue can be resolved easily by setting `len(unit_params)` be 0, 1, `len(tiling_dims)` w.r.t. no factor(1), shared factor, corresponding factors. 

Considering above issues, I propose another approach inspired from [helion's tile](https://helionlang.com/api/generated/helion.language.tile.html), replacing `tiling_dims` and `unit_params` with:

- **`shapes`**: a sequence of full shapes 
- **`tiling_dims`**: a list of int/tuples as tiling dimensions (can't be zero length) 

and some optional paramters are related to shared memory usage, such as whether loading two tiles seperately or simultaneously, generating different formulas when estimating total size. I haven't thoroughly throught about naming and how to generalize the memory formula, we can default to one of them.


### Some examples 

#### RoPE 

Args:

- **`shapes`**: `([n_q_head, hd], [n_kv_head, hd])`
	- we can pass original shapes and let tiling strategy function handle padding
	- compatible with padded shapes as well
- **`tiling_dims`**: `(0, 0)`
	- `n_q_head` and `n_kv_head` can be broken up into tiles
	- the rest of dimensions are padded to next power of 2 values
	- equivalent to current `num_params = (pad_hd,)` 

- keeping other parameters: `safety_margin`, `dtype_size`, `memory_multiplier`

Return:
- tuple of final tiling sizes: `([block_size_q, pad_hd], [block_size_kv, pad_hd])`
	- same structure as **`shapes`**
	


#### A simplified scenario for matmul, BLOCK_K is pre-defined 

Args:

- **`shapes`**: `([M, BLOCK_K], [BLOCK_K, N])`
- **`tiling_dims`**: `(0, 1)` 
	- `M` and `N` can be broken up into tiles

Return:
- tuple of final tiling sizes: `([BLOCK_M, BLOCK_K], [BLOCK_K, BLOCK_N])`
	- same structure as **`shapes`**

#### A more complicated scenario for matmul

Args:
- **`shapes`**: `([M, K], [K, N])`
- **`tiling_dims`**: `((0, 1), (0, 1))` 
	- all dimensions can be broken up into tiles
	- or passing (`None`, `None`) to indicate shape [M, K] and shape [K, N] are all sliceable. *(Not sure if it's a good idea)*

Return:
- tuple of final tiling sizes: `([BLOCK_M, BLOCK_K], [BLOCK_K, BLOCK_N])`
	- same structure as **`shapes`**

This case would better combine with some extra parameters (strides?) for coalesced loading, identifying contiguous dimension, and so on to achieve better performance. Not required in our very first version.

What do you think about this approach?

noemotiovon commented:
Hi @Tcc0403, We've updated the code to use `Tuple[Tuple[int, ...], ...]` throughout, including:
- Function signatures and return types
- All operator calls (ROPE and GEGLU)
- Documentation and examples

The new interface is indeed cleaner and more consistent with PyTorch conventions. Thanks for catching this!

Tcc0403 commented on src/liger_kernel/ops/backends/_ascend/ascend-ub-manager-design.md:
````diff
@@ -0,0 +1,462 @@
+│ Helper function to normalize        │
+│ tiling_dim (int or tuple) to set    │
+└─────────────────────────────────────┘
+```
````
minor formatting issue: fit texts in boxes

noemotiovon replied:
Done

Tcc0403 commented:
That was fast! Overall LGTM, requesting reviews in case I overlook something.

hipudding commented on src/liger_kernel/ops/backends/_ascend/ub_manager.py:
```diff
@@ -0,0 +1,341 @@
+        # Validate tiling dimensions are within shape bounds
+        if not tiling_dim_set or any(dim_idx < 0 or dim_idx >= len(shape) for dim_idx in tiling_dim_set):
+            # Invalid tiling_dim, use conservative default
+            max_safe_sizes.append(1)
```
max_safe_sizes.append(_DEFAULT_UB_CAPACITIES * 0.8)

noemotiovon replied:
This branch handles fallback behavior for invalid parameters. When tiling_dims or shape does not meet the expected constraints, it is no longer possible to accurately derive a reasonable block_size based on UB capacity, so the logic must degrade gracefully.
There are two main fallback options in this situation:
1. Make block_size sufficiently small (e.g., set it to 1) to maximize memory safety and ensure that the kernel can still run correctly.
2. Avoid meaningful block partitioning (effectively using a very large block size). This may lead to UB overflow, but given that the parameters are already invalid, a reliable block size cannot be computed anyway.

noemotiovon replied:
Done.

hipudding commented on src/liger_kernel/ops/backends/_ascend/ops/geglu.py:
```diff
@@ -0,0 +1,244 @@
+    # Always use the unified NPU kernel
+    # When adjusted_block_size >= n_cols, the loop executes only once (no tiling)
+    # When adjusted_block_size < n_cols, the loop handles tiling automatically
+    _geglu_tanh_forward_kernel_npu[(n_rows,)](
```
We may not be able to directly reuse the GPU approach here. Since the number of NPU cores is far smaller than that of GPUs, launching too many threads can lead to severe scheduling latency issues and potentially very poor performance.

In addition, each thread may access only a small amount of data, which can result in very low utilization of the Unified Buffer (UB).

I suggest setting the number of threads to match the number of NPU cores (configurable), and partitioning elements based on the following principle:
The minimum granularity should be computation-related data. For example, for element-wise operations, the minimum granularity is an element; for operations that require full-row computation, the minimum granularity is one row. Then determine how many such minimum-granularity elements can fit into the UB.

In short, the UB should be filled as much as possible, and the number of threads should be aligned with the number of cores. This helps reduce DMA transfer instructions and core scheduling overhead, thereby maximizing the utilization of the compute units.
